File: decision_tree.py
'''
Created on Nov 30st, 2019
Tree-Based Regression Methods
@author: Shihao Liu
'''

import numpy as np
import argparse

parser = argparse.ArgumentParser()
parser.add_argument('-maximum_height', default=20, type=int)
parser.add_argument('-tree_mode', default='cart_regression', type=str)

args = parser.parse_args()

class DecisionTree:
    def __init__(self, is_leaf=False, left=None, right=None, height=1, split_feature=None,
                 split_value=None, leaf_nums=0, threshold=1):
        self.left = left
        self.right = right
        self.height = height
        self.is_leaf = is_leaf
        self.split_feature = split_feature
        self.split_value = split_value
        self.leaf_nums = leaf_nums
        self.threshold = threshold

    def build_tree(self, tree, dataset):
        best_split_feature, split_value = self.choose_best_feature(tree, dataset)
        tree.split_feature = best_split_feature
        tree.split_value = split_value
        if tree.is_leaf is True or best_split_feature is None:
            return
        dataset_1, dataset_2 = self.split_dataset(dataset, best_split_feature, split_value)
        tree.left = DecisionTree(height=tree.height+1)
        tree.right = DecisionTree(height=tree.height+1)

        self.build_tree(tree.left, dataset_1)
        self.build_tree(tree.right, dataset_2)

    def split_dataset(self, dataset, feature, value):
        dataset_1 = dataset[np.nonzero(dataset[:, feature] <= value)[0]]
        dataset_2 = dataset[np.nonzero(dataset[:, feature] > value)[0]]
        return dataset_1, dataset_2

    def choose_best_feature(self, tree, dataset):
        if len(set(dataset[:, -1])) == 1 or tree.height >= args.maximum_height:
            tree.is_leaf = True
            return None, np.mean(dataset[-1])
        maximum_gain = 0
        feature_nums = len(dataset[0]) - 1 # 最后一列默认为label
        loss = self.compute_loss(dataset, args.tree_mode)
        for f in range(feature_nums):
            for val in dataset[:, f]:
                dataset_1, dataset_2 = self.split_dataset(dataset, f, val)
                loss_1 = self.compute_loss(dataset_1, args.tree_mode)
                loss_2 = self.compute_loss(dataset_2, args.tree_mode)
                new_loss = loss_1 + loss_2
                gain = loss - new_loss
                if gain < tree.threshold:
                    continue
                if gain > maximum_gain:
                    best_split_feature, split_value = f, val
                    maximum_gain = gain
                    self.loss = new_loss
        if maximum_gain == 0 or maximum_gain < tree.threshold:
            tree.is_leaf = True
            return None, np.mean(dataset[-1])
        else:
            return best_split_feature, split_value

    def compute_loss(self, dataset, mode):
        if mode == 'cart_regression':
            return np.var(dataset[:, -1]) * np.shape(dataset)[0]

    def predict(self, tree, feature):
        if tree.is_leaf:
            return tree.split_value
        if feature[tree.split_feature] <= tree.split_value:
            return self.predict(tree.left, feature)
        else:
            return self.predict(tree.right, feature)
            
    def tree_illustration(self, tree):
        queue = [tree]
        while queue:
            tree = queue.pop(0)
            if tree.is_leaf or not type(tree.split_feature) == int:
                print(tree.split_value)
                continue
            print(tree.split_feature, tree.split_value)
            queue.extend([tree.left, tree.right])

        return

class Data:
    def __init__(self, file_path):
        self.file_path = file_path

    def load_file(self):
        dataset = np.loadtxt(self.file_path)
        return dataset

if __name__ == '__main__':
    data = Data('./bikeSpeedVsIq_train.txt')
    training_set = data.load_file()
    cart = DecisionTree()
    cart.build_tree(cart, training_set)

    # cart.tree_illustration(cart)

    data = Data('./bikeSpeedVsIq_test.txt')
    test_set = data.load_file()
    loss = 0
    for example in test_set:
        prediction = cart.predict(cart, example)
        loss += (prediction - example[-1]) ** 2

    print("模型在test_set上的最终MSE值为{:.2f}".format(loss))

File: decision_tree.cpp
#include<iostream>
#include<vector>
#include<cmath>
#include<unordered_set>

#include "data.h"

using namespace std;

class SplitResults{
    public:
    int feature;
    float value;
};

class SplitData{
    public:
    Data subdata_1;
    Data subdata_2;
};

class DecisionTree
{
    public:
    DecisionTree *left;
    DecisionTree *right;
    int height;
    int maximum_height = 15;
    bool is_leaf;
    int split_feature;
    float split_value;
    float threshold;

    // DecisionTree(): left(NULL), right(NULL), height(1), is_leaf(false), split_feature(NULL), split_value(NULL) {};
    DecisionTree(
                 int h=1,
                 DecisionTree *l = NULL, 
                 DecisionTree *r = NULL, 
                 bool is_leaf = false,
                 int feat = -1,
                 float val = 0,
                 float t = 1
                 ):
                 left(l), right(r), height(h), is_leaf(is_leaf), split_feature(feat), split_value(val), threshold(t) {};
    
    void build_tree(DecisionTree* tree, vector<vector<float>>& dataset)
    {
        if (dataset.size() == 0)
        {
            return;
        }
        SplitResults feature_and_val = choose_best_feature(tree, dataset);
        tree -> split_feature = feature_and_val.feature;
        tree -> split_value = feature_and_val.value;
        tree -> height += 1;
        if (tree -> is_leaf == true || tree -> split_feature == -1)
        {
            return;
        }
        SplitData new_dataset = split_dataset(dataset, tree -> split_feature, tree -> split_value);
        vector<vector<float>> dataset_1 = new_dataset.subdata_1;
        vector<vector<float>> dataset_2 = new_dataset.subdata_2;

        tree -> left = new DecisionTree(tree -> height + 1);
        tree -> right = new DecisionTree(tree -> height + 1);

        build_tree(tree -> left, dataset_1);
        build_tree(tree -> right, dataset_2);
        return;
    }

    SplitResults choose_best_feature(DecisionTree* tree, vector<vector<float>>& dataset){
        SplitResults result;
        bool flag = true;
        float value = dataset[0].back();
        for (size_t i=0; i < dataset.size(); ++i)
        {
            if (dataset[i].back() != value)
            {
                flag = false;
                break;
            }
        }
        if (flag == true || tree -> height >= tree -> maximum_height)
        {
            result.feature = -1;
            result.value = compute_mean(dataset);
            tree -> is_leaf = true;
            return result;
        }
        float loss = compute_loss(dataset);
        float maximum_gain = 0;
        for (size_t i = 0; i < dataset[0].size() - 1; ++i) // 最后一列是label
        {
            for (size_t j = 0; j < dataset.size(); ++j)
            {
                SplitData new_dataset = split_dataset(dataset, i, dataset[j][i]);
                vector<vector<float>> dataset_1 = new_dataset.subdata_1;
                vector<vector<float>> dataset_2 = new_dataset.subdata_2;
                float new_loss = (compute_loss(dataset_1) + compute_loss(dataset_2)) / 2;
                float gain = loss - new_loss;
                if (gain < tree -> threshold)
                {
                    continue;
                }
                if (gain > maximum_gain)
                {
                    result.feature = i;
                    result.value = dataset[j][i];
                    maximum_gain = gain;
                }
            }
        }
        if (maximum_gain == 0 || maximum_gain < tree -> threshold)
        {
            result.feature = -1;
            result.value = compute_mean(dataset);
            tree -> is_leaf = true;
        }
        return result;
    }

    SplitData split_dataset(vector<vector<float>>& dataset, int feature, float value){
        SplitData new_dataset;
        for (size_t i = 0; i < dataset.size(); ++i)
        {
            if (dataset[i][feature] <= value)
            {
                new_dataset.subdata_1.push_back(dataset[i]);
            }
            else
            {
                new_dataset.subdata_2.push_back(dataset[i]);
            }
        }
        return new_dataset;
    }

    float compute_mean(Data& dataset){
        float total = 0;
        for (size_t i = 0; i < dataset.size(); ++i)
        {
            total += dataset[i].back();
        }
        return total / dataset.size();
    }

    float compute_loss(Data& dataset){
        float loss = 0;
        float mean = compute_mean(dataset);
        for (size_t i = 0; i < dataset.size(); ++i)
        {
            loss += (dataset[i].back() - mean) * (dataset[i].back() - mean);
        }
        loss = sqrt(loss / dataset.size());
        return loss;
    }

    float predict(DecisionTree* tree, vector<float>& dataset)
    {
        if (tree -> is_leaf == true || tree -> split_feature == -1)
        {
            return tree -> split_value;
        }
        if (dataset[tree -> split_feature] <= tree -> split_value)
        {
            return predict(tree -> left, dataset);
        }
        else
        {
            return predict(tree -> right, dataset);
        }
    }
};

int main()
{
    Data training_set = LoadData("/Users/liushihao/Desktop/搜索引擎基础/GBDT/bikeSpeedVsIq_train.txt");
    DecisionTree tree;
    tree.build_tree(&tree, training_set);

    Data test_set = LoadData("/Users/liushihao/Desktop/搜索引擎基础/GBDT/bikeSpeedVsIq_test.txt");
    for (size_t i = 0; i < test_set.size(); ++i)
    {   
        cout << test_set[i].back() << '\t' << tree.predict(&tree, test_set[i]) << endl;
    }
    return 0;
}